src/mavsdk/plugins/component_metadata/component_metadata_impl.cpp

```diff
@@ -0,0 +1,559 @@
+{
+    LogDebug() << "ComponentMetadataImpl::retrieve_metadata for compid " << static_cast<int>(compid)
+               << ", type " << static_cast<int>(type);
+
+    auto& component = _mavlink_components[compid].components[type];
+    bool crc_valid;
+    uint32_t crc;
+    bool is_translation;
+    std::string uri;
+
+    if (component.get_next_uri(crc_valid, crc, is_translation, uri)) {
+        // If translation locale is not set, skip translations
+        if (is_translation && !_translation_locale) {
+            retrieve_metadata(compid, type);
+            return;
+        }
+
+        std::optional<std::filesystem::path> cached_file_option{};
+        std::string file_cache_tag{};
+        if (_file_cache && crc_valid) {
+            file_cache_tag = get_file_cache_tag(compid, type, crc, is_translation);
+            cached_file_option = _file_cache->access(file_cache_tag);
+        }
+
+        if (cached_file_option) {
+            LogDebug() << "Using cached file " << cached_file_option.value();
+            component.current_metadata_path() = cached_file_option.value();
+            retrieve_metadata(compid, type);
+        } else {
+            LogDebug() << "Downloading json " << uri;
+            std::string download_path;
+            uint8_t target_compid = compid;
+
+            if (uri_is_MAVLinkFTP(uri, download_path, target_compid)) {
+                // Create compid-specific tmp subdir, to ensure multiple components don't overwrite
+                // files from each other
+                const std::filesystem::path tmp_download_path =
+                    _tmp_download_path / std::to_string(compid);
+                if (!std::filesystem::exists(tmp_download_path)) {
+                    std::error_code err;
+                    std::filesystem::create_directory(tmp_download_path, err);
+                }
+
+                const std::filesystem::path local_path =
+                    tmp_download_path / std::filesystem::path(download_path).filename();
+                _system_impl->mavlink_ftp_client().download_async(
+                    download_path,
+                    tmp_download_path,
+                    true,
+                    [this, &component, local_path, compid, type, file_cache_tag](
+                        MavlinkFtpClient::ClientResult download_result,
+                        MavlinkFtpClient::ProgressData progress_data) {
+                        if (download_result == MavlinkFtpClient::ClientResult::Next) {
+                            LogDebug()
+                                << "File download progress: " << progress_data.bytes_transferred
+                                << '/' << progress_data.total_bytes;
+                            // TODO: detect slow link (e.g. telemetry), and cancel download
+                            // (fallback to http) e.g. by estimating the remaining download time,
+                            // and cancel if >40s
```

Comment:
How do you know there is a http fallback?

Reply:
You can check if the uri is set. It's not guaranteed, but PX4 generally always sets a http fallback uri.